### DISCUSSION TITLE
Help in finding ADC maps

### DISCUSSION THREAD

**Discussion by evrtyq** [dipy/dipy#3064] [Reconstruction]
Hi, all

I am newbie to dMRi. So replying with helpful resources would be very useful. Most of my trial and errors is credited to this link https://carpentries-incubator.github.io/SDC-BIDS-dMRI/aio/index.html. I am working of **PROSTATE DW Images**

 I have some 4D nifti images. I understand that it consists of 2 volumes.( `len(image_data.shape) ==4` and `image_data.shape[-1] ==2`). The .bval file has 2 b-values which I assume corresponds to each of the volume. Given below is the code I used.

```
import nibabel as nib
from dipy.io.gradients import read_bvals_bvecs
from dipy.core.gradients import gradient_table
import dipy.reconst.dti as dti
from dipy.segment.mask import threshold_multiotsu

img = nib.load("nifti_file_obtained_from_using_dicom2nifti.nii.gz")
img_data = np.asanyarray(img.dataobj)

img_data.shape  #(256,256,64,2)

bval = "bval_file_obtained_from_using_dicom2nifti.bval"
bvec = "bvec_file_obtained_from_using_dicom2nifti.bvec"
gt_bvals, gt_bvecs = read_bvals_bvecs(bval, bvec)

 #threshold = threshold_multiotsu(img_data , classes = 2)
#I need to create a mask but everywhere there is only how to create a brainmask and not a prostate mask

dti_model = dti.TensorModel(gtab)
dti_fit = dti_model.fit(img_data)
adc_img = dti_fit.md  #The link I mentioned in the begining says mean diffusivity(md) is also known as ADC
nib.save(nib.Nifti1image(adc_img, img.affine),"path_to_adc.nii.gz")
```

The problem is that the final ADC image has only 2 unique values(0 and 1e-6) for b-val (0,0). Is this right?

https://github.com/loli/medpy/blob/master/notebooks/scripts/medpy_apparent_diffusion_coefficient.py.ipynb helped me in creating ADC maps and they are completely different from what I got. But the images I have from which I can retrieve the ADC map are:

N - bval = 0, 0      -      Both shouldn't be b0 image
Y - bval = 0, 2000    -      did work
N - bval = 50, 400,1000       -     no b0 image and 3 b-values present instead of 2
N - bval = 50, 500, 1000       -     no b0 image and 3 b-values present instead of 2
N - bval = 50, 1400       -     no b0 image

I think only the 2nd one was successful using medpy. Any help or direction towards resources would be very helpful. Thanks.

**Comment by pjsjongsung** [comment]
DTI has 6 degrees of freedom, which means it needs at least 6 volumes (better if more) to fit the matrix.
As you have shown below, it also needs at least one b0 image as well.

I am not sure why you are trying different sets of b values, but could I suggest trying to fit the model with all the b values instead of 2? (i.e. 0, 50, 400, 500, 1000, 1400, 2000)

**Reply by evrtyq** [reply]
@pjsjongsung  Thanks for the fast reply

> I am not sure why you are trying different sets of b values, but could I suggest trying to fit the model with all the b values instead of 2?

I have many 4D images and each b-value pair or trio corresponds to that image. So far I have only been able to find the b-value of the single subject with b-value(0,2000)

> DTI has 6 degrees of freedom, which means it needs at least 6 volumes (better if more) to fit the matrix.

Do you suggest that I use a different reconstruction method because for most of the individual cases I have a maximum of 2 or 3 volumes.
If I had more than 6 volumes, is my method of retrieving the ADC map using dipy right?

**Reply by pjsjongsung** [reply]
I don't think there are any models that can work with 3 volumes. However, the code seems right and following @arokem 's comments might help you with your problem.

**Comment by arokem** [comment]
Hello @evrtyq! Though @pjsjongsung is correct that DTI requires at least six different measurements, you might still be able to calculate the MD/ADC (these are both the same thing for your purpose) from these data, even if you have just two measurements, with the caveat that this will be the ADC in just one direction of diffusion. For this, you do need two different measurements with two different b-values, though. What do you see when you execute: `print(gt_bvals)`? Are the files `bval_file_obtained_from_using_dicom2nifti.nii.gz` and `bvec_file_obtained_from_using_dicom2nifti.nii.gz` text files? They should be. If you open them up in a text editor, what do you see in them?

**Reply by evrtyq** [reply]
@arokem 
> Are the files `bval_file_obtained_from_using_dicom2nifti.nii.gz` and `bvec_file_obtained_from_using_dicom2nifti.nii.gz` text files? They should be. If you open them up in a text editor, what do you see in them?

I am sorry for the typo. I have updated the post. They are text files. 


> What do you see when you execute: `print(gt_bvals)`?

I get the numpy_array with values:( Please note that these are a sample values of different studies that I came across. This is not the case for a single subject)

1. 0 0
2. 0 2000
3. 50 400 1000
4. 50 500 1000
5. 50 1400

**Comment by evrtyq** [comment]
@arokem Do you suggest that it is normal when I get an image with only 2 unique values (0 and 1e-6) in my ADC image?

**Reply by arokem** [reply]
No. This seems like an error to me.

But I can't really help you until you answer my questions above, so I am still waiting.

As a reminder, my questions were:

> ...  If you open them [bval/bvec files] up in a text editor, what do you see in them?

And: 

> What do you see when you execute: print(gt_bvals)?
